app/services/report_summariser.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+class ReportSummariser
+  def initialize(group_column, report_rows, report)
+    @group_column = group_column
+    @report_rows = report_rows
+    @report = report
+  end
+
+  def call
+    insert_summary_rows unless exclude_summaries?
+
+    @report_rows
+  end
+
+  private
+
+  def insert_summary_rows
+    grouped_rows = []
+    previous_grouping = nil
+
+    @report_rows.each_with_index do |row, row_index|
+      current_grouping = row[@group_column]
+
+      if previous_grouping.present? && current_grouping != previous_grouping
+        grouped_rows << build_summary_row(@group_column, previous_grouping, @report.summary_row)
+      end
+
+      grouped_rows << row
+
+      if last_row?(row_index)
+        grouped_rows << build_summary_row(@group_column, previous_grouping, @report.summary_row)
+      end
+
+      previous_grouping = current_grouping
+    end
+
+    @report_rows = grouped_rows
+  end
```

Comment:
IMO the moment we make some processing/aggregation in Ruby is when things get slow to a crawl and get worse as more requirements pile up. The way I see it, we should move all the burden to the DB.

There's much more we can benefit from PostgreSQL. It has many advanced analytical capabilities and specifically [Window Functions](https://www.postgresql.org/docs/9.6/tutorial-window.html) I think should provide us with all we need.

Reply:
Would all the reports queries have to be done in SQL if we did this though? Or how would that work nicely with things like Ransack and ActiveRecord scopes?

Reply:
Did I mention it's already twice as fast as our current implementation? :trollface:

Reply:
sure, but I would also like the leverage the feaures PostgreSQL so we maintain less code.